gammapy/modeling/models/spectral.py

```diff
@@ -47,34 +46,45 @@ def __sub__(self, model):
     def __rsub__(self, model):
         return self.__sub__(model)
 
-    def _evaluate_gradient(self, energy, eps=1e-12):
-        x = self.parameters.values
-        frozen = np.array([_.frozen for _ in self.parameters])
+    def _evaluate_gradient(self, energy, eps):
+        n = len(self.parameters)
+        f = self(energy)
+        shape = (n, len(np.atleast_1d(energy)))
+        df_dp = np.zeros(shape)
 
-        def func(xk):
-            kwargs = {par.name: val * par.unit for val, par in zip(xk, self.parameters)}
-            kwargs = self._convert_evaluate_unit(kwargs, energy)
-            return self.evaluate(energy, **kwargs).value
+        for idx, parameter in enumerate(self.parameters):
+            if parameter.frozen or eps[idx] == 0:
+                continue
 
-        grad = approx_jacobian(x=x, func=func, epsilon=eps)
-        grad[:, frozen] = 0
-        return grad.T
+            parameter.value += eps[idx]
+            df = self(energy) - f
+            df_dp[idx] = df.value / eps[idx]
 
-    def evaluate_error(self, energy):
+            # Reset model to original parameter
+            parameter.value -= eps[idx]
+
+        return df_dp
+
+    def evaluate_error(self, energy, precision=1e-4):
         """Evaluate spectral model with error propagation.
 
         Parameters
         ----------
         energy : `~astropy.units.Quantity`
             Energy at which to evaluate
+        precision : float
+            Numerical precision of the gradient evaluation.
```

Comment:
Suggest to rename the parameter and description. It's not the precision of the gradient. It's the step size for the parameter.

Reply:
I renamed the parameter to `epsilon` (that's what we have for `SpectralModel.spectral_index()` as well) and changed the description accordingly.